bodo/utils/utils.py

```diff
@@ -1882,3 +1878,53 @@ class AWSCredentials:
     secret_key: str
     session_token: str | None = None
     region: str | None = None
+
+
+def create_arg_hash(*args, **kwargs):
+    """
+    Create a hash encompassing the string representations of all the args and kwargs.
+    This is typically used to generate a unique and repeatable function name.
+    Args:
+        args and kwargs: the variables that have some effect on the contents of the function.
+    """
+    concat_str_args = "".join(map(str, args)) + "".join(
+        f"{k}={v}" for k, v in kwargs.items()
+    )
+    arg_hash = hashlib.sha256(concat_str_args.encode("utf-8"))
+    return arg_hash.hexdigest()
+
+
+def bodo_exec(func_text, glbls, loc_vars, real_globals):
+    """
+    Take a string containing a dynamically generated function with a given name and exec
+    it into existence and make the resulting function Numba cacheable.
+    Args:
+        func_text: the text of the new function to be created
+        glbls: the globals to be passed to exec
+        loc_vars: the local var dict to be passed to exec
+        real_globals: should be passed globals() from the calling scope
+    """
+    # Get hash of function text.
+    text_hash = hashlib.sha256(func_text.encode("utf-8")).hexdigest()
+    # Use a regular expression to find and add hash to the function name.
+    pattern = r"(def\s+)(\w+)(\s*\()"
+    found_pattern = re.search(pattern, func_text)
+    assert found_pattern
```

Comment:
Minor: just adding an error message for potentially easier debugging.
```suggestion
    assert found_pattern, "bodo_exec: function definition not found"
```